feat(adapters): Support additional headers for OpenAI backend

planetf1 opened the pull request:
<!--
Thank you for your pull request. Please review and complete the sections below.
-->

### Which issue(s) does this pull-request address?

<!--
Please include a link to an issue in the tracker.  The issue describes the problem to be solved.  If there is no issue raised for this PR then either raise one with a summary and description of the problem or add a summary and description of the problem here
-->

Closes: #532

### Description

[ Updated with revised format / variable ]

<!-- Provide a description of the change, pay special attention to describing any breaking changes.  The description describes the resolution to the problem described in the linked issue (or to the problem outlined in this PR). -->

Adds support for LiteLLM provider to add additional headers
```
export OPENAI_EXTRA_HEADERS="MY_API_KEY=ab592c9f3e001b"
```

To test (for example, with OpenAI, also set:
```
OPENAI_API_KEY=my-api-key
OPENAI_CHAT_MODEL=my-model
OPENAI_API_BASE=https://my-endpoint/
```

Tested ok

### Checklist

<!-- For completed items, change [ ] to [x]. -->

- [x] I have read the [contributor guide](https://github.com/i-am-bee/beeai-framework/blob/main/CONTRIBUTING.md)
- [x] I have [signed off](https://github.com/i-am-bee/beeai-framework/blob/main/CONTRIBUTING.md#developer-certificate-of-origin-dco) on my commit
- [x] Linting passes: `yarn lint` or `yarn lint:fix`
- [x] Formatting is applied: `yarn format` or `yarn format:fix`
- [x] Unit tests pass: `yarn test:unit`
- [ ] E2E tests pass: `yarn test:e2e`
- [ ] Tests are included <!-- Bug fixes and new features should include tests -->
- [x] Documentation is changed or added
- [x] Commit messages and PR title follow [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary)

vabarbosa commented on python/beeai_framework/adapters/openai/backend/chat.py:
```diff
@@ -30,6 +31,21 @@ def provider_id(self) -> ProviderName:
 
+        # Extra headers -- ignored if invalid
+        extra_headers = None
+        extra_headers_json = _settings.get("extra_headers", os.getenv("OPENAI_EXTRA_HEADERS"))
```
`extra_headers` is supported across multiple services, e.g.,
- https://docs.litellm.ai/docs/providers/anthropic#passing-extra-headers-to-anthropic-api
- https://docs.litellm.ai/docs/providers/bedrock#passing-extra-headers--custom-api-endpoints

to that end should this really be `OPENAI_EXTRA_HEADERS` and should it be in openai adapter? i would think we want this in the LiteLLM adapter so it could be applicable across other adapters (and replace the `OPENAI_` prefix with something more generic)

planetf1 replied:
Yes, I'd agree - if we can do this at the litellm level it will be much more flexible

planetf1 replied:
Now moved to `LITELLM_EXTRA_HEADERS` and implemented in the litellm adapter

vabarbosa commented on python/beeai_framework/adapters/openai/backend/chat.py:
```diff
@@ -30,6 +31,21 @@ def provider_id(self) -> ProviderName:
+                pass
+
+        if extra_headers:
+            _settings["headers"] = extra_headers
```
LiteLLM also accepts `extra_headers` params: 
![image](https://github.com/user-attachments/assets/7590a1af-73da-4244-a876-eb205f40aca9)

so do we have to copy it over to `headers`? we end up with both `headers` and `extra_headers` in the `_settings`

planetf1 replied:
it probably should be extra_headers so that we just augment what is already there?

planetf1 replied:
Yes, I'd agree - if we can do this at the litellm level it will be much more flexible

planetf1 replied:
Reviewing litellm docs/search : summary:
```
headers: These are the standard HTTP headers that litellm uses for authentication and content negotiation. For example, litellm automatically sets headers like Content-Type: application/json and Accept: application/json. You can also use the headers parameter to override or add to these standard headers. A common use case is to set an Authorization header with your API key if needed for certain model providers.

extra_headers:  This parameter is designed for passing custom headers that are not part of the standard HTTP headers used by litellm.  extra_headers is particularly useful when you are routing requests through an API management gateway or a reverse proxy and need to include specific headers for authorization, project identification, or other custom logic handled by the gateway. The value of extra_headers should be a JSON string representing the headers you want to add.
```
For this use case the latter seems more appropriate - so updated accordingly

vabarbosa reviewed (changes requested):
thanks... just a couple comments. your thoughts are appreciated

planetf1 commented:
I can look at this tomorrow to refactor to make more generic. (if urgent ok to merge than update after)

planetf1 commented:
Rather than adding multiple environment variables and managing precedence across inheritance, are we ok to just add a single one?

Previously we had `OPENAI_EXTRA_HEADERS` (in the above working code)

If this is set at the LiteLLM adapter level it might be `LIGHTLLM_EXTRA_HEADERS`

It might be more generic in future if we use `BEEAI_EXTRA_HEADERS` or `BEEAI_LLM_EXTRA_HEADERS` ... ? But the implementation would then differ, and potentially be incomplete.

My thought is to just have a single variable, to name it `BEEAI_LLM_EXTRA_HEADERS`, but only to implement it in LiteLLM adapter

planetf1 commented:
Moved to draft whilst refactoring/testing. Will move back to ready once all comments are acted on and tests complete.

planetf1 commented on python/beeai_framework/adapters/openai/backend/chat.py:
```diff
@@ -30,6 +31,21 @@ def provider_id(self) -> ProviderName:
+                if isinstance(parsed_headers, dict):
+                    extra_headers = parsed_headers
+            except json.JSONDecodeError:
+                pass
```
Given the other points about moving to litellm (which many other (all?) adapters are based on currently I think I'd defer further refactor until we identify an additional need for this outside litellm

planetf1 commented:
Ready for review

Tomas2D commented on python/beeai_framework/adapters/litellm/chat.py:
```diff
@@ -55,14 +58,39 @@ class LiteLLMChatModel(ChatModel, ABC):
+
+        if extra_headers_json:
+            try:
+                parsed_headers = json.loads(extra_headers_json)
+                if isinstance(parsed_headers, dict):
+                    extra_headers = parsed_headers
```
The parsing does not seem correct. In `TS` the input is not JSON, but a string with commas (KEY=VALUE,KEY2=VALUE2).

planetf1 replied:
Agreed, in that it was intentionally written to be a json string. As you point out this differs from typescript.

Headers can get complex, and need escaping if we want to handle the most general case, so going with a well defined format (like json) would seem to be more appropriate?

planetf1 replied:
it will also mean the env var will get set to the same value as the extra_headers property used by the ChatModel.

planetf1 replied:
In team call 20250312 consensus was that the extra headers should use simple strings, not json - will update

planetf1 replied:
* Updated implementation to use key value pairs.
* Tested with rits
* Error checking/parsing is simple, just checking for '='. Any issues will be met with a log entry, and ignoring the value set.
* No change to how extra_headers property is used. This remains in JSON format.

Tomas2D reviewed (commented):
If your project uses multiple LLM providers, then having just `LITELLM_EXTRA_HEADERS` would not work. Let's be provider-specific.

planetf1 commented:
Updated implementation to:
 - use common util to manage headers
 - added unit test for utility
 - have a provider specific OPENAI_EXTRA_HEADERS
 - updated docs
 - tested against internal inference service
 
 This can be applied to the other litellm based adapters. I suggest we complete this PR first (and merge when approved). The others can be added in a subsequent PR
